File: web/src/lib/stats.ts
// Aggregates.
//
// Doc 0: "Aggregates are never stored, only queried." Every number below is derived on demand
// from a ScoutEvent[]; nothing here is persisted or cached.

import type { Alliance, EventType, Phase, ScoutEvent } from '../contracts';
import { phaseBounds, pointValuesArePlaceholders, type SeasonConfig } from '../season';

export interface ScoreBreakdown { red: number; blue: number }

export interface TeamStats {
  team: number;
  alliance: Alliance | null;
  shotAttempts: number;
  shotsMade: number;
  /** Made / attempted. Null when the team never attempted a shot. */
  accuracy: number | null;
  reloads: number;
  /**
   * Doc 0's vocabulary: "the interval between one `reload` event and the next `reload` event
   * for the same team. Acquire to acquire, not acquire to score, so a missed shot still costs
   * a cycle... An unterminated final cycle is discarded, not counted."
   */
  cycleCount: number;
  medianCycleSeconds: number | null;
  bestCycleSeconds: number | null;
  cycleSeconds: number[];
  /** Doc 1: "Shot rate: interval between consecutive shot events." */
  avgShotIntervalSeconds: number | null;
  defenseSeconds: number;
  immobileSeconds: number;
  fouls: number;
  pointsContributed: number;
  lowConfidenceEvents: number;
  byPhase: Record<Phase, { attempts: number; made: number }>;
}

const EMPTY_PHASES = (): Record<Phase, { attempts: number; made: number }> => ({
  auto: { attempts: 0, made: 0 },
  teleop: { attempts: 0, made: 0 },
  endgame: { attempts: 0, made: 0 },
  unknown: { attempts: 0, made: 0 },
});

function median(xs: number[]): number | null {
  if (xs.length === 0) return null;
  const s = [...xs].sort((a, b) => a - b);
  const m = s.length >> 1;
  return s.length % 2 ? s[m] : (s[m - 1] + s[m]) / 2;
}

function mean(xs: number[]): number | null {
  return xs.length ? xs.reduce((a, b) => a + b, 0) / xs.length : null;
}

/**
 * Points for one made shot in a phase, at a given goal.
 *
 * v3 added `goal`, so this no longer has to assume every shot went in the high goal. A null
 * goal means the model could not place the shot -- it scores 0 rather than guessing, which
 * keeps the accuracy comparison honest about what the pipeline actually knows.
 *
 * All point values are zero placeholders until the game is public; doc 0: "Do not invent
 * values to make a test pass."
 */
export function pointsFor(phase: Phase, goal: string | null, cfg: SeasonConfig | null): number {
  if (!cfg || !goal) return 0;
  const group = cfg.pointValues[phase];
  if (!group) return 0;
  return group[`shot_made_${goal}`] ?? 0;
}

function pairedSeconds(
  events: ScoutEvent[],
  startType: EventType,
  endType: EventType,
  matchEnd: number
): number {
  let total = 0;
  let openedAt: number | null = null;
  for (const e of events) {
    if (e.eventType === startType && openedAt === null) openedAt = e.tSeconds;
    else if (e.eventType === endType && openedAt !== null) {
      total += Math.max(0, e.tSeconds - openedAt);
      openedAt = null;
    }
  }
  if (openedAt !== null) total += Math.max(0, matchEnd - openedAt);
  return total;
}

export function computeTeamStats(
  team: number,
  allEvents: ScoutEvent[],
  alliance: Alliance | null = null,
  cfg: SeasonConfig | null = null,
  lowConfidenceThreshold = 0.5
): TeamStats {
  const matchEnd = cfg ? phaseBounds(cfg).matchEnd : Infinity;
  const mine = allEvents.filter((e) => e.team === team).sort((a, b) => a.tSeconds - b.tSeconds);

  const byPhase = EMPTY_PHASES();
  let shotAttempts = 0;
  let shotsMade = 0;
  let reloads = 0;
  let fouls = 0;
  let pointsContributed = 0;
  let lowConfidenceEvents = 0;
  const reloadTimes: number[] = [];
  const shotTimes: number[] = [];

  for (const e of mine) {
    if (e.confidence < lowConfidenceThreshold) lowConfidenceEvents++;
    switch (e.eventType) {
      case 'shot_attempt':
        shotAttempts++;
        shotTimes.push(e.tSeconds);
        byPhase[e.phase].attempts++;
        break;
      case 'shot_made':
        shotsMade++;
        byPhase[e.phase].made++;
        pointsContributed += pointsFor(e.phase, e.goal, cfg);
        break;
      case 'reload':
        reloads++;
        reloadTimes.push(e.tSeconds);
        break;
      case 'foul':
        fouls++;
        break;
      default:
        break;
    }
  }

  // Acquire-to-acquire. An unterminated final cycle is simply never formed, since a cycle
  // only exists between two reloads.
  const cycleSeconds: number[] = [];
  for (let i = 1; i < reloadTimes.length; i++) {
    cycleSeconds.push(reloadTimes[i] - reloadTimes[i - 1]);
  }
  const shotIntervals: number[] = [];
  for (let i = 1; i < shotTimes.length; i++) shotIntervals.push(shotTimes[i] - shotTimes[i - 1]);

  return {
    team,
    alliance,
    shotAttempts,
    shotsMade,
    accuracy: shotAttempts > 0 ? shotsMade / shotAttempts : null,
    reloads,
    cycleCount: cycleSeconds.length,
    medianCycleSeconds: median(cycleSeconds),
    bestCycleSeconds: cycleSeconds.length ? Math.min(...cycleSeconds) : null,
    cycleSeconds,
    avgShotIntervalSeconds: mean(shotIntervals),
    defenseSeconds: pairedSeconds(mine, 'defense_start', 'defense_end', matchEnd),
    immobileSeconds: pairedSeconds(mine, 'immobile_start', 'immobile_end', matchEnd),
    fouls,
    pointsContributed,
    lowConfidenceEvents,
    byPhase,
  };
}

/**
 * Reconstruct the match score from events, for the "reconstructed vs TBA" indicator.
 *
 * Events with a null team contribute nothing -- an unidentified robot cannot be credited to
 * an alliance, and guessing would quietly inflate the accuracy number this exists to test.
 */
export function reconstructScore(
  events: ScoutEvent[],
  alliances: { red: number[]; blue: number[] } | null,
  cfg: SeasonConfig | null
): ScoreBreakdown {
  const out: ScoreBreakdown = { red: 0, blue: 0 };
  if (!alliances || !cfg) return out;
  const sideOf = (team: number | null): Alliance | null => {
    if (team == null) return null;
    if (alliances.red.includes(team)) return 'red';
    if (alliances.blue.includes(team)) return 'blue';
    return null;
  };
  for (const e of events) {
    const side = sideOf(e.team);
    if (!side) continue;
    if (e.eventType === 'shot_made') out[side] += pointsFor(e.phase, e.goal, cfg);
  }
  return out;
}

/** Whether a reconstructed score means anything yet. */
export function scoringIsMeaningful(cfg: SeasonConfig | null): boolean {
  return cfg != null && !pointValuesArePlaceholders(cfg);
}

File: ingest/stats.py
"""Aggregates and score reconstruction, SCHEMA_VERSION 3.

Doc 0: "Aggregates are never stored, only queried. If a stat is needed often enough to hurt,
add a materialized view, not a column." Everything here is computed on demand from event rows.

Cycle time is doc 0's vocabulary definition: the interval between one `reload` and the next
`reload` for the same TEAM. Acquire to acquire, so a missed shot still costs a cycle. Per team
rather than per track, because track ids are job-local and a re-identified robot may span
several. An unterminated final cycle is discarded, not counted.
"""

import json
import statistics
from functools import lru_cache
from pathlib import Path

SEASONS_DIR = Path(__file__).resolve().parent.parent / "contracts" / "seasons"


@lru_cache(maxsize=8)
def season_config(year: int) -> dict | None:
    """Doc 0: selected by the job's `season`, so old footage stays analyzable.

    Returns None for an unknown season rather than falling back to a current one -- doc 0:
    "Anything unrecognized is a bug, not a fallback."
    """
    path = SEASONS_DIR / f"{year}.json"
    if not path.exists():
        return None
    with open(path, "r", encoding="utf-8") as handle:
        return json.load(handle)


def points_for(phase: str, goal: str | None, cfg: dict | None) -> int:
    """Points for one made shot in a phase, at a given goal.

    v3 added `goal`, so this no longer assumes every shot went in the high goal. A null goal
    means the model could not place the shot: it scores 0 rather than guessing, which keeps
    the accuracy comparison honest about what the pipeline actually knows.

    Every value is a zero placeholder until the game is public; doc 0: "Do not invent values
    to make a test pass."
    """
    if not cfg or not goal:
        return 0
    group = cfg.get("point_values", {}).get(phase) or {}
    return int(group.get(f"shot_made_{goal}", 0))


def scoring_is_meaningful(cfg: dict | None) -> bool:
    """False while every point value is still a zero placeholder."""
    if not cfg:
        return False
    groups = cfg.get("point_values", {}).values()
    return any(v for group in groups for v in (group or {}).values())


def legal_goals(cfg: dict | None) -> set[str] | None:
    """Legal `goal` values for a season, or None when the season is unknown."""
    if not cfg:
        return None
    return set(cfg.get("goals") or [])


def _alliance_of(team, alliances) -> str | None:
    if team is None or not alliances:
        return None
    if team in (alliances.get("red") or []):
        return "red"
    if team in (alliances.get("blue") or []):
        return "blue"
    return None


def reconstruct_score(events: list[dict], alliances: dict | None, cfg: dict | None) -> dict:
    """Sum made shots into an alliance score.

    An event with team=None contributes nothing: an unidentified robot cannot be credited to
    an alliance, and guessing would quietly inflate the number this exists to test.
    """
    score = {"red": 0, "blue": 0}
    if not alliances:
        return score
    for event in events:
        side = _alliance_of(event.get("team"), alliances)
        if side is None:
            continue
        if event.get("event_type") == "shot_made":
            score[side] += points_for(
                event.get("phase", "unknown"), event.get("goal"), cfg
            )
    return score


def _paired_seconds(events, start_type: str, end_type: str, match_end: float) -> float:
    """Total seconds spanned by start/end pairs; an unclosed interval runs to match_end."""
    total = 0.0
    opened_at = None
    for event in sorted(events, key=lambda e: e.get("t_seconds") or 0.0):
        if event.get("event_type") == start_type and opened_at is None:
            opened_at = event.get("t_seconds") or 0.0
        elif event.get("event_type") == end_type and opened_at is not None:
            total += max(0.0, (event.get("t_seconds") or 0.0) - opened_at)
            opened_at = None
    if opened_at is not None:
        total += max(0.0, match_end - opened_at)
    return total


def team_stats(
    team: int,
    events: list[dict],
    cfg: dict | None,
    event_key: str | None,
    matches_played: int,
    min_confidence: float = 0.0,
    low_confidence_threshold: float = 0.5,
) -> dict:
    """Contract E's team_stats shape. Fields may be added additively; none renamed."""
    if cfg:
        match_end = float(cfg["auto_seconds"]) + float(cfg["teleop_seconds"])
    else:
        match_end = float("inf")

    mine = [e for e in events if e.get("team") == team]

    reload_times = sorted(
        (e.get("t_seconds") or 0.0) for e in mine if e.get("event_type") == "reload"
    )
    cycles = [reload_times[i] - reload_times[i - 1] for i in range(1, len(reload_times))]

    shot_times = sorted(
        (e.get("t_seconds") or 0.0) for e in mine if e.get("event_type") == "shot_attempt"
    )
    intervals = [shot_times[i] - shot_times[i - 1] for i in range(1, len(shot_times))]

    logged_attempts = sum(1 for e in mine if e.get("event_type") == "shot_attempt")
    made = sum(1 for e in mine if e.get("event_type") == "shot_made")
    # A robot cannot score more shots than it took. The two are separate events -- the golden
    # fixture has 80 attempts and 55 makes, none sharing a timestamp -- so a scout logging under
    # time pressure can easily record the makes and miss the attempts. Believing that literally
    # reports "1 made of 0 attempted, accuracy unknown", which reads as a broken tool rather than
    # as missing input. Taking the larger of the two is a floor on what was attempted, and it can
    # never push accuracy above 1.
    attempts = max(logged_attempts, made)

    return {
        "team": team,
        "event_key": event_key,
        "min_confidence": min_confidence,
        "matches_played": matches_played,
        "cycles": len(cycles),
        # Median, not mean: one immobile robot produces a single enormous interval.
        "avg_cycle_seconds": statistics.median(cycles) if cycles else None,
        "shot_attempts": attempts,
        "shots_made": made,
        "shot_accuracy": (made / attempts) if attempts else None,
        "avg_shot_interval_seconds": statistics.fmean(intervals) if intervals else None,
        "reloads": len(reload_times),
        "defense_seconds": _paired_seconds(mine, "defense_start", "defense_end", match_end),
        "immobile_seconds": _paired_seconds(mine, "immobile_start", "immobile_end", match_end),
        "fouls": sum(1 for e in mine if e.get("event_type") == "foul"),
        "low_confidence_events": sum(
            1 for e in mine if (e.get("confidence") or 0.0) < low_confidence_threshold
        ),
    }
